### DISCUSSION TITLE
Help getting Metals working as the LSP for Scala in Helix?

### DISCUSSION THREAD

**Discussion by csgordon** [helix-editor/helix#9517] [Q&A] [answered]
Hi, I'm using Helix for editing Rust and LaTeX, and things generally just work once the appropriate LSP is on the path. But with Scala and Metals, I'm not having such luck. Most of what I find online about Scala and Metals suggests it should work as I was expecting, but so far doesn't work for me. I've got Metals installed, and in my PATH, and it looks like Helix finds it:
```
$ hx --health | grep scala
scala              ✓ metals           None               ✓                  ✘                  ✓
```
But no LSP functionality works: no hover support, no project build... can anyone suggest something I might be missing?

Thanks in advance!

**Comment by filipwiech** [comment]
Hey! :slightly_smiling_face: 

AFAIK it's most likely because you have to import the build first, before Metals will start providing any real support for the project. Normally (e.g. in VS Code, nvim-metals, etc.) Metals will send an LSP message asking whether to import the build after each project definition change (for example in `build.sbt`) where user can send a response back. Unfortunately Helix doesn't support the required API endpoint at the moment, so the request is ignored and Metals never starts. I think the PR required to make this work is this one: #5492. There's also another nice-to-have so that status updates would be shown: #5535. :+1: 

Some solution might be do what is described [here on the sbt blog](https://eed3si9n.com/helix-and-scala/) and import the project using another editor first and then switch to Helix. There's also an [in progress PR](https://github.com/scalameta/metals/pull/6064) in Metals itself to automatically import the build without any action from the user which could help as well (not sure if it will work for the initial fresh import or only for the further updates though). :+1: 

Hope it helps, would be great to see Helix support the required LSP requests out of the box. :slightly_smiling_face:

**Comment by csgordon** [comment]
Thanks for the response. This high level explanation makes a lot of sense (that there's a setup configuration missing), and the referenced PRs on helix and metals certainly seem related.

Unfortunately I had already attempted the instructions on Eugene's blog, and they don't seem to work for me; Helix still doesn't really interact with Metals. For example, ```<Space>-k``` over an identifier just reports ```No configured language server supports hover```.

Part of me is wondering now, though, after looking at those PRs to Helix, if I've got some subtle misconfiguration of Helix such that it's not talking to Metals properly. ```:lsp-workspace-command``` gives me ```No active language servers for this document support workspace commands```, which seems odd given some screenshots in those threads (though maybe those are also dependent on the unmerged PRs?).

**Comment by jpaju** [comment, marked as answer]
I've been able to import Scala projects in Helix by first opening a Scala/BST file, then running `:lsp-workspace-command build-import`. If this is the first import for the specific project, it might take a while and no progress is being reported in the Helix UI. 

Sometimes Metals might get stuck and requires running `:lsp-workspace-command cascade-compile`. As a last resort you may need to remove .metals, .bsp and .bloop folders if nothing else helps.

**Reply by csgordon** [reply]
Hm, weird, maybe I've borked my setup somehow? When I open a Scala file in the project and try ```:lsp-workspace-command build-import``` it doesn't look like it does nothing, it actively fails, with
```
No active language servers for this document support workspace commands
```

Yet ```hx --health scala``` reports
```
Configured language servers:
  ✓ metals: /usr/local/bin/metals
Configured debug adapter: None
Highlight queries: ✓
Textobject queries: ✘
Indent queries: ✓
```

**Reply by jpaju** [reply]
If you just type `:lsp-workspace-command `, does Helix suggest/autocomplete some commands?

**Reply by csgordon** [reply]
Earlier I did get a list of suggestions, but when I just tried now to get a couple examples to include, I got nothing.  Same terminal, no environment changes.

**Reply by jpaju** [reply]
I'm guessing there is something wrong with your metals installation because I cannot replicate the behavior you're describing. If you start helix in debug mode (`hx <scala-file> -vv`), open the LSP workspace picker and check the logs(`:log-open`), how does it look like?

**Reply by csgordon** [reply]
Thank you! This was helpful. The log file showed a bunch of internal server errors from metals with threadpool-related exceptions, which is a clear indicator of a metals issue. After a bit more digging, I discovered that with the current Homebrew-installed metals, the version reported by metals does not match the version Homebrew thinks is installed (though the installed version does claim to support everything I need). I don't know what exactly is broken there, but reordering my PATH so that Coursier's version of metals is found does give me meaningful pop-ups for ```:lsp-workspace-command```. I've run the ```build-import``` and ```cascade-compile``` subcommands, and now I at least get hover information for Scala built-ins in the standard library, which is a vast improvement!  Still no hover results for things from my own project, though, even in the same file.

**Reply by csgordon** [reply]
Okay, I also get hover info for dependencies pulled in from build.sbt. It's only my own code that I lack information for. This seems like purely a metals issue now, rather than Helix
